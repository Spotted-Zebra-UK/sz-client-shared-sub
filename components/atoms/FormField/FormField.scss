@import '../../../variables';

.FormField {
  &__Label__Text {
    font-size: $p2-font-size;
    line-height: $small-line-height;
    text-align: left;
  }

  &__Label__Text--HasError {
    color: $purple;
  }

  &__Error {
    color: $purple;
    font-size: $p2-font-size;
    min-height: 16px;
    line-height: $small-line-height;
    margin-top: 8px;
    text-align: left;
  }
}
