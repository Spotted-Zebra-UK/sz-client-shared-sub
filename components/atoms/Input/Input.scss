@import '../../../variables';

.Input {
  border: 0;
  border-bottom: 2px solid $black;
  font-size: $p1-font-size;
  height: 36px;
  line-height: $medium-line-height;
  width: 100%;

  &:disabled {
    background-color: transparent;
  }
}
