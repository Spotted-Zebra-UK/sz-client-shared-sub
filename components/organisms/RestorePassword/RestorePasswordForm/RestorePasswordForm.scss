@import '../../../../variables';

.RestorePasswordForm {
  .FormField--email {
    margin-bottom: 8px;
  }

  &__SubmitButton {
    margin: 0 auto;
  }
}
