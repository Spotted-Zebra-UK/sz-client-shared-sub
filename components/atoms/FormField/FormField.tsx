import './FormField.scss';
import React, { FunctionComponent } from 'react';

interface IFormField {
  error?: string;
  fieldName: string;
  label?: string;
}

const FormField: FunctionComponent<IFormField> = ({
  children,
  error,
  label,
  fieldName,
}) => {
  return (
    <div className={`FormField FormField--${fieldName}`}>
      <label className="FormField__Label" htmlFor={fieldName}>
        {label ? (
          <div
            className={`FormField__Label__Text${
              error ? ' FormField__Label__Text--HasError' : ''
            }`}
          >
            {label}
          </div>
        ) : null}
        {children}
        <div className="FormField__Error">{error}</div>
      </label>
    </div>
  );
};

export default FormField;
