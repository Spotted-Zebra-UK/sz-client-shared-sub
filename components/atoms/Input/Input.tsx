import './Input.scss';
import React, { FunctionComponent } from 'react';

interface IInput {
  className?: string;
  name: string;
  onChange: (value: string, name: string) => void;
  placeholder?: string;
  value: string;
  type?: 'text' | 'password' | 'date' | 'number';
  isDisabled?: boolean;
  id?: string;
}

const Input: FunctionComponent<IInput> = ({
  className,
  name,
  onChange,
  placeholder,
  value,
  type,
  isDisabled,
  id,
}) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value, name);
  };

  return (
    <input
      className={`Input${className ? ` ${className}` : ''}`}
      name={name}
      onChange={handleChange}
      onBlur={handleChange}
      value={value}
      type={type || 'text'}
      disabled={isDisabled || false}
      placeholder={placeholder}
      id={id}
    />
  );
};

export default Input;
