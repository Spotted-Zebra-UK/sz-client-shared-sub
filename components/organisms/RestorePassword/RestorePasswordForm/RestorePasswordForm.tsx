import './RestorePasswordForm.scss';
import React, { FC, useState } from 'react';
import validate from '../../../../helpers/validate';
import FormField from '../../../atoms/FormField/FormField';
import Input from '../../../atoms/Input/Input';
import SquareButton from '../../../molecules/SquareButton/SquareButton';

interface IRestorePasswordFormValues {
  email: string;
}

interface IRestorePasswordFormErrors {
  email: string[];
}

interface IRestorePasswordFormForm {
  onSubmit: (email: string) => void;
}

const RestorePasswordForm: FC<IRestorePasswordFormForm> = props => {
  const [values, setValues] = useState<IRestorePasswordFormValues>({
    email: '',
  });
  const [errors, setErrors] = useState<
    IRestorePasswordFormErrors | undefined
  >();

  const restorePasswordFormValidate = () => {
    return validate(values, {
      email: {
        email: {
          message: '^Not a valid email',
        },
      },
    });
  };

  const handleChange = (value: string, name: string) => {
    setValues(prevValues => ({ ...prevValues, [name]: value }));
    setErrors(prevErrors =>
      prevErrors ? { ...prevErrors, [name]: [] } : prevErrors
    );
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const newErrors = restorePasswordFormValidate();
    if (newErrors) {
      setErrors(newErrors);
    } else {
      props.onSubmit(values.email);
    }
  };

  return (
    <form className="RestorePasswordForm" onSubmit={handleSubmit}>
      <div className="RestorePasswordForm__Fields">
        <FormField
          error={errors && errors.email && errors.email.join(' ')}
          fieldName="email"
          label="Email"
          isLabelVisible={!!values.email}
        >
          <Input
            name="email"
            onChange={handleChange}
            placeholder="Email"
            value={values.email}
          />
        </FormField>
      </div>
      <SquareButton type="submit">Reset password</SquareButton>
    </form>
  );
};

export default RestorePasswordForm;
