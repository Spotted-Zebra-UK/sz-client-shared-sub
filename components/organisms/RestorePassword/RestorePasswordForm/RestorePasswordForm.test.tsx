import { shallow } from 'enzyme';
import FormField from '../../../atoms/FormField/FormField';
import Input from '../../../atoms/Input/Input';
import RestorePasswordForm from './RestorePasswordForm';

describe('RestorePasswordForm', () => {
  let onSubmit: (email: string) => void;

  beforeEach(() => {
    onSubmit = jest.fn();
  });

  // it('should render correctly by default', () => {
  //   const wrapper = shallow(<RestorePasswordForm onSubmit={onSubmit} />);

  //   expect(toJson(wrapper)).toMatchSnapshot();
  //   expect(wrapper.find(FormField).props().error).toBeUndefined();
  //   expect(wrapper.find(Input).props().value).toBe('');
  // });

  it('should change input value on input change', () => {
    const emailValue = 'test@email.com';
    const wrapper = shallow(<RestorePasswordForm onSubmit={onSubmit} />);

    wrapper.find(Input).props().onChange(emailValue, 'email');
    expect(wrapper.find(Input).props().value).toBe(emailValue);
  });

  it('should call provided onSubmit', () => {
    const emailValue = 'test@email.com';
    const wrapper = shallow(<RestorePasswordForm onSubmit={onSubmit} />);

    wrapper.find(Input).props().onChange(emailValue, 'email');
    expect(wrapper.find(Input).props().value).toBe(emailValue);

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(eventMock.preventDefault).toHaveBeenCalled();
    expect(onSubmit).toHaveBeenCalledWith(emailValue);
  });

  it('should show error and not call onSubmit if submit invalid email', () => {
    const emailValue = 'invalid-email';
    const wrapper = shallow(<RestorePasswordForm onSubmit={onSubmit} />);

    wrapper.find(Input).props().onChange(emailValue, 'email');

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(eventMock.preventDefault).toHaveBeenCalled();
    expect(wrapper.find(FormField).props().error).toBe('Not a valid email');
    expect(onSubmit).not.toHaveBeenCalled();
  });

  it('should hide error if error is visible and input changed', () => {
    const validEmailValue = 'test@email.com';
    const invalidEmailValue = 'test-emailValue';
    const wrapper = shallow(<RestorePasswordForm onSubmit={onSubmit} />);

    wrapper.find(Input).props().onChange(invalidEmailValue, 'email');

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(wrapper.find(FormField).props().error).toBe('Not a valid email');

    wrapper.find(Input).props().onChange(validEmailValue, 'email');
    expect(wrapper.find(FormField).props().error).toBe('');
  });
});
