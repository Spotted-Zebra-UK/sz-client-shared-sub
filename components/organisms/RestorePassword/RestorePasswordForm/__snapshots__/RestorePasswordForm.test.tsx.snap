// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestorePasswordForm should render correctly by default 1`] = `
<form
  className="RestorePasswordForm"
  onSubmit={[Function]}
>
  <div
    className="RestorePasswordForm__Fields"
  >
    <FormField
      fieldName="email"
      label="Email"
    >
      <Input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        value=""
      />
    </FormField>
  </div>
  <SquareButton
    type="submit"
  >
    Reset password
  </SquareButton>
</form>
`;
